### ISSUE TITLE
Communicate when TXT cleanup is OK & group domains please

### ISSUE THREAD

**Issue by wbt** [certbot/certbot#9602]
When using a DNS challenge to renew a cert, there is a prompt like this for each domain name on the Cert:

>Please deploy a DNS TXT record under the name
_acme-challenge.example.org with the following value:
ABCDEfghi89....(random string)
Before continuing, verify the record is deployed.
(This must be set up in addition to the previous challenges; do not remove,
replace, or undo the previous challenge tasks yet. Note that you might be
asked to create multiple distinct TXT records with the same name. This is
permitted by DNS standards.)
>\- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
>Press Enter to Continue

A few usability issues are present here, which maybe could be split out into separate Issue filings if desired:
1) The message for the first domain in a comma-separated list doesn't show the parenthetical part; you just have to remember that from the last renewal until after completing the step and proceeding to the second one.  It should be displayed for the first domain too.
2) The message says "do not remove, replace, or undo the previous challenge tasks yet" but never tells the user when they *can* remove prior challenges.  Given the short duration of certs, this can lead to TXT records piling up and feel like tech debt.  Perhaps the "IMPORTANT NOTES" displayed upon success could be extended to say "you can now remove TXT records from prior challenges, other than the ones you just added" if that's accurate. 
3) The order in which the challenges are presented makes things much more difficult than needed, especially given that two separate DNS records have to be added to the same domain for `example.org` and `*.example.org` but you have to switch multiple times through the .com, .net, and whatever other domains before cycling back through to get the second record for the asterisk-subdomain for the first domain.  It would be better if only one TXT record could be used instead of two for such a use case, or if the interface somehow indicated which one of the two domain patterns would be validated by that record, but at least clustering them together would be helpful.  Perhaps in order of entry or in lexical order of the domain strings *reversed* might be easy?  Sorting in lexical order with multiple levels of sort for each level of the domain starting from the end would probably be best, but just reversing the strings and ordering by that would be a great improvement over the status quo.
4) Related to the last one, the text states that the challenge value must be in place before continuing to the next domain, but it's not clear if that's actually true or if it's OK if the challenges are in place by the end of the set.  It might be better to group the strings together with a message like this:

>Please deploy a DNS TXT record under the name in the first column with the value shown in the second: 
_acme-challenge.example.org           ABCDEfghi89....(random string)
_acme-challenge.example.com          WXYZabcde2....(random string)
Before continuing, verify the records are deployed.
(Each of these must be set up in addition to the previous challenges; do not remove,
replace, or undo the previous challenge tasks yet. Note that you might be
asked to create multiple distinct TXT records with the same name. This is
permitted by DNS standards.)
>\- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
>Press Enter to Continue

Showing some extra space between those columns isn't easily supported in GitHub but I would increase it a bit.

Thanks for your work on this application!

**Comment by osirisinferi** [comment]
Hi @wbt, thanks for the issue!

While I agree with your number 2 point, I'm not sure if I understand the first one. The fact that the parenthical part about not removing previous TXT RRs was deliberately: TXT RRs from *previous* Certbot runs are not required any longer, only the TXT RRs from the current run. So I don't fully understand what you mean by "you just have to remember that from the last renewal" as those TXT RRs are useless for the current renewal.

With regard to your third point: perhaps Certbot shouldn't do any sorting at all, but just keep the order the user provided the hostnames. But I'm not sure if that's technically possible with how Certbot handles the hostnames currently. Maybe some reverse sorting might be in order indeed.

Point 4: I believe Certbot triggers the validation at the end indeed. However, it *might* be more useful for the user to get them step by step, so one doesn't forget one by accident. But I'm interested in hearing the opinion on the Certbot team.

**Comment by wbt** [comment]
> TXT RRs from _previous_ Certbot runs are not required any longer, only the TXT RRs from the current run.

The text is misleading then; what I understood from it is that these records from previous Certbot runs must remain in place through at least the next renewal.  

For point 4, the inconvenience of having to change back and forth between domains and revisit the same domain again seems much more common and weighty than what's associated with the risk of forgetting one.  You could add some UI allowing the user to move up and down and hit space to mark one as completed to overcome that, or just do what happens when one is missed: there's a failure and a need to do it again, maybe requiring a bit of a wait depending on the record's TTL.

**Comment by osirisinferi** [comment]
> The text is misleading then

Not sure if it's *misleading*. The whole reason why it isn't mentioned with the first challenge is because *that* would be confusing. While it's not *explicit*, one can assume the message relates to the previous challenge Certbot *just* provided, not from previous runs. Personally, I'm a little bit hesitant to add more text to the already long message to be honest.

> the success message really should say something like "you may now remove the TXT records added during this process."

Probably not only the success message, as failed challenges can't be reused either. So if the Certbot team would like to add an additional message about removal of the TXT RRs, it would also be necessary at an error.

> You could add some UI allowing the user to move up and down and hit space to mark one as completed to overcome that

Certbot *used* to have an UI based on (n)curses, but that was dropped at some point in time (quite early I believe). Currently Certbot does not have such an interactive UI unfortunately, so this wouldn't be possible.

It should be possible to provide all the challenges at one go of course. But I'm *guessing* there went some thought into the current way of providing the challenges, so I'm curious to hear what the Certbot team has to say about it :slightly_smiling_face:

**Comment by wbt** [comment]
>While it's not explicit, one can assume the message relates to the previous challenge Certbot *just* provided, not from previous runs. 

I don't think that's generally a safe assumption that someone with no knowledge of how it works should make.  Other protocols require DNS challenge strings to be left in place indefinitely, and Certbot interface text focuses on how important it is to leave records in place saying nothing about when they can be cleaned up, so it's not that unreasonable to assume they have to be left there just like with other protocols in order to avoid Cert cancellation/invalidation. Stating explicitly that removal is OK, on success and on error, would be a valuable improvement.

**Comment by alexzorin** [comment]
Let me preface this comment by saying that using `--manual` without hooks really is a last resort and not the focus of Certbot.

>The message for the first domain in a comma-separated list doesn't show the parenthetical part; you just have to remember that from the last renewal until after completing the step and proceeding to the second one. It should be displayed for the first domain too.

Based on prior discussion in this issue, the point looks to be moot.

>but never tells the user when they can remove prior challenges

Perhaps a minor wording change can be applied here, something like: "*do not remove, replace, or undo the previous challenge tasks until after the certificate is issued.*"

>order of the domain strings reversed might be easy

Sure, OK, happy to review a PR for this. Please sort within `._internal.plugins.manual.Authenticator.perform` and only do so if it's a request without a `--manual-auth-hook`.

> It might be better to group the strings together with a message like this:

I'm open to this change as well. 

Myself, I'd probably prefer all the records upfront instead of being dripfed. However, the overall effect this would have on the UX is unclear and I'm not saying that we'd definitely merge this. The change needs to all make sense for a set of HTTP or mixed HTTP/DNS  challenges. Anybody who is interested in writing a draft PR, I'd be willing to take a look.

**Comment by wbt** [comment]
> "do not remove, replace, or undo the previous challenge tasks until after the certificate is issued."

That would seem even more clear that the records from the last renewal have to remain in place - which seems to be the opposite of the intended message?  How about:

> "do not remove, replace, or undo other TXT records added during this same renewal attempt until after the certificate is issued."

**Comment by osirisinferi** [comment]
It's not just after issuance, if a challenge has been triggered, there are two scenarios: it either succeeded and the challenge won't need to be attempted for 30 days (in case of Let's Encrypt, valid challenges are cached for 30 days), so the TXT RR can be removed *OR* it failed and the TXT RR won't be useful any longer, because the challenge has been invalidated entirely.

So unless Certbot crashes or the user exits Certbot while it was running somehow (Ctrl-C), all the challenge tokens will be useless after Certbot exits, whether the certificates was issued or not. With that in mind, I'd suggest something like:

> do not remove, replace, or undo these challenge tasks until after Certbot has exited

or something similar, replacing "the previous" with "these" so it refers to the challenges from *this* run (implicitely, I admit) and replace issuance with exit.

@wbt Note that Certbot can mix `http-01` challenges and `dns-01` challenges (as stated earlier), so specifying TXT RRs in the message is not always correct.

**Comment by wbt** [comment]
Combined with a reminder at the end, that sounds like an improvement.  Maybe instead of "these challenge tasks" it might be clearer to say "the challenge records from this run of Certbot".  

For the table proposed in point 4, I'd be OK having two tables with manual advancement from one to the other, where one table is the DNS challenges and the other is HTTP challenges, or otherwise have HTTP challenges handled separately.
